source/extensions/tracers/opencensus/opencensus_tracer_impl.cc

```diff
@@ -0,0 +1,248 @@
+      }
+      break;
+    }
+
+    case OpenCensusConfig::grpc_trace_bin: {
+      const Http::HeaderEntry* header =
+          request_headers.get(Http::LowerCaseString{"grpc-trace-bin"});
+      if (header != nullptr) {
+        found = true;
+        parent_ctx = ::opencensus::trace::propagation::FromGrpcTraceBinHeader(
+            header->value().getStringView());
+      }
+      break;
+    }
+
+    case OpenCensusConfig::cloud_trace_context: {
+      const Http::HeaderEntry* header =
+          request_headers.get(Http::LowerCaseString{"x-cloud-trace-context"});
+      if (header != nullptr) {
+        found = true;
+        parent_ctx = ::opencensus::trace::propagation::FromCloudTraceContextHeader(
+            header->value().getStringView());
+      }
+      break;
+    }
+    }
+    // First header found wins.
+    if (found) {
+      break;
+    }
+  }
+
+  // Honor Envoy's tracing decision.
+  ::opencensus::trace::AlwaysSampler always_sampler;
+  ::opencensus::trace::NeverSampler never_sampler;
+  // This is safe because opts are not used after StartSpan.
+  ::opencensus::trace::StartSpanOptions opts{&never_sampler};
+  if (traced) {
+    opts.sampler = &always_sampler;
+  }
+
+  if (parent_ctx.IsValid()) {
+    return ::opencensus::trace::Span::StartSpanWithRemoteParent(name, parent_ctx, opts);
+  }
+  return ::opencensus::trace::Span::StartSpan(name, /*parent=*/nullptr, opts);
+}
+
+Span::Span(const Tracing::Config& config,
+           const envoy::config::trace::v2::OpenCensusConfig* oc_config,
+           Http::HeaderMap& request_headers, const std::string& operation_name,
+           SystemTime /*start_time*/, const Tracing::Decision tracing_decision)
+    : span_(StartSpanHelper(operation_name, tracing_decision.traced, request_headers, *oc_config)),
+      oc_config_(oc_config) {
+  span_.AddAttribute("OperationName", config.operationName() == Tracing::OperationName::Ingress
+                                          ? "Ingress"
+                                          : "Egress");
+}
+
+Span::Span(const envoy::config::trace::v2::OpenCensusConfig* oc_config,
+           ::opencensus::trace::Span&& span)
+    : span_(std::move(span)), oc_config_(oc_config) {}
+
+void Span::setOperation(absl::string_view operation) {
+  span_.AddAnnotation("setOperation", {{"operation", operation}});
+}
+
+void Span::setTag(absl::string_view name, absl::string_view value) {
+  span_.AddAttribute(name, value);
+}
+
+void Span::log(SystemTime /*timestamp*/, const std::string& event) {
+  // timestamp is ignored.
+  span_.AddAnnotation(event);
+}
+
+void Span::finishSpan() { span_.End(); }
+
+void Span::injectContext(Http::HeaderMap& request_headers) {
+  using OpenCensusConfig = envoy::config::trace::v2::OpenCensusConfig;
+  for (const auto& outgoing : oc_config_->outgoing_trace_context()) {
+    switch (outgoing) {
+    case OpenCensusConfig::trace_context:
+      request_headers.addCopy(
+          Http::LowerCaseString{"traceparent"},
```

Comment:
extract these keys to a singleton, and use `addReferenceKey`. See how `ZipkinCoreConstants` works.

Reply:
Done. Thanks!